### PR TITLE
sender MUST limit bursts, SHOULD limit to IW

### LINKED ISSUE #3892: MUST requirement on bursts smaller than IW is too restrictive
"Implementations MUST either use pacing or another method to limit such bursts to the initial congestion window; see Section 7.2."  

This precludes use of segmentation offload of sizes greater than IW. Hardware does not pace TSO for TCP or USO for UDP. In datacenters we routinely send bursts that are higher without causing loss. Both Linux and Windows TCP send bursts larger than IW (up to 64k) at higher bandwidths. Higher bursts are required for achieving high data rates. 

This text should be changed to a SHOULD.

### PR DESCRIPTION
**Pull request by janaiyengar**
Closes #3892.
Also does a bit of cleanup.

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -877,12 +877,15 @@ similar to a sender's response on a Retransmission Timeout (RTO) in TCP
...
-to the initial congestion window; see {{initial-cwnd}}.
+A sender SHOULD pace sending of all in-flight packets based on input from the
+congestion controller. Sending multiple packets into the network without any
+delay between them creates a packet burst that might cause short-term congestion
+and losses.
+
+A sender SHOULD either use pacing or another method to limit such bursts to the
+initial congestion window; see {{initial-cwnd}}. A sender might have prior
+knowledge that the network path to the receiver has deep buffers, and might
+therefore allow larger bursts.
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:888]:
```suggestion
initial congestion window; see {{initial-cwnd}}. A sender that has prior
knowledge that the network path to the receiver has deep buffers might allow
larger bursts.
```

Comment (ianswett) [reply draft-ietf-quic-recovery.md:888]: Yes, it's certainly design.

**Comment by janaiyengar** [comment]
@ianswett : Can you provide some text? Also, shall we make this part of the consensus call?

**Comment by janaiyengar** [comment]
@ianswett : I don't think we are missing it. The PR says that you SHOULD pace or limit the burst size to IW at all times. Isn't that what you want -- to limit bursts to IW?

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -877,12 +877,15 @@ similar to a sender's response on a Retransmission Timeout (RTO) in TCP
...
-them creates a packet burst that might cause short-term congestion and losses.
-Implementations MUST either use pacing or another method to limit such bursts
-to the initial congestion window; see {{initial-cwnd}}.
+A sender SHOULD pace sending of all in-flight packets based on input from the
+congestion controller.
+
+Sending multiple packets into the network without any delay between them creates
+a packet burst that might cause short-term congestion and losses. Senders MUST
+either use pacing or limit such bursts. Senders SHOULD limit bursts to the
+initial congestion window; see {{initial-cwnd}}. A sender with prior knowledge
```
Comment (ianswett) [inline draft-ietf-quic-recovery.md:886]:
```suggestion
initial congestion window; see {{initial-cwnd}}. A sender with knowledge
```

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -877,12 +877,14 @@ similar to a sender's response on a Retransmission Timeout (RTO) in TCP
...
-them creates a packet burst that might cause short-term congestion and losses.
-Implementations MUST either use pacing or another method to limit such bursts
-to the initial congestion window; see {{initial-cwnd}}.
+A sender SHOULD pace sending of all in-flight packets based on input from the
+congestion controller.
+
+Sending multiple packets into the network without any delay between them creates
+a packet burst that might cause short-term congestion and losses. Senders MUST
+either use pacing or limit such bursts. Senders SHOULD limit bursts to the
+initial congestion window; see {{initial-cwnd}}. A sender with knowledge that
```
Comment (pravb) [inline draft-ietf-quic-recovery.md:886]: I don't think we should refer to deep buffers. At high data rates in DC the path can process packets at multi-Gbps line rate so even with reasonable buffer sizes the bursts can be absorbed.  How about:

"A sender with knowledge that the network path to the receiver can absorb larger bursts may use a higher limit."

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:886]: Done.

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -877,12 +877,15 @@ similar to a sender's response on a Retransmission Timeout (RTO) in TCP
...
+the network path to the receiver can absorb larger bursts may use a higher
```
Comment (ianswett) [inline draft-ietf-quic-recovery.md:887]:
```suggestion
the network path to the receiver can absorb larger bursts MAY use a higher
```

Or change it to a can.

**Pull request merged**